### ISSUE TITLE
ERROR failed to open repository: unable to create format manager: invalid repository password

### ISSUE THREAD

**Issue by elcid2012** [kopia/kopia#4177]
Pretty easy bug to reproduce, and results in kopia completely failing to start.

1. From a fresh kopia docker container, spin up and login successfully. 
2. Connect to an S3-compatible repository (minio) which has no issues reading/writing information to the bucket.
3. Set the repository password to the same environment variable `KOPIA_PASSWORD`
4. Restart kopia with a simple `docker restart kopia` - this also happens when the host is rebooted
5. Get this error, which prevents kopia from successfully starting:

```
ERROR failed to open repository: unable to create format manager: invalid repository password
ERROR unable to initialize repository: unable to initialize repository: error opening repository: unable to open repository: unable to create format manager: invalid repository password
```

Kopia continues in a reboot loop if `restart: always` is configured in the docker command (in my case, using docker-compose).

As long as kopia is never restarted or the host is never rebooted, it will continue to chug along and write snapshots to the repository without any failures.

**Comment by deuts** [comment]
This is my issue as well. Every time, without fail, when restarting the Docker container, it throws this same error.

**Comment by deuts** [comment]
@elcid2012 I think I might have found a solution to my problem. Would it be possible that you have special characters in your `KOPIA_PASSWORD`? In my case I had a `$` character, but when doubling it like so `$$`, it solved my problem.

**Comment by elcid2012** [comment]
@deuts 

Nope... I just tested it again with a non-complex password "P3pper61" which is the same password used for both the web interface and server password variables. Rebooted the container and got the same error message. Thanks for the suggestion! I was pretty sure I was already using a simple password, but this exercise confirmed it.

**Comment by TilKenneth** [comment]
I just had this error, it was a wrong password. I am not sure, but do you use env set
1. `KOPIA_PASSWORD=P3pper61`
2. `KOPIA_PASSWORD="P3pper61"`

My senses tell me, at least on batch (cmd), the quotes become part of the password if put in a cmd file and quoted. While on command line they are interpreted as expected either way (ie: `--password="P3pper61"` or `--password "P3pper61"`).

I can't test them right now, because I'm restoring everything I put into my repo, to make sure it actually restores... Bad case of multiple parallel snapshot from cron, beause my script didnt use a "if script is running, just return"-mechanism.

**Comment by elcid2012** [comment]
> 1. `KOPIA_PASSWORD=P3pper61`
> 2. `KOPIA_PASSWORD="P3pper61"`

@TilKenneth You are on to something, because that's exactly the issue. In the documentation, the docker-compose example uses a combination of "-" and non-"-" deliniations, and those are compiled differently. See https://kopia.io/docs/installation/#docker-images

It has:
```
        environment:
            # Set repository password
            KOPIA_PASSWORD: "SECRET"
            USER: "User"
```

Whereas I was using
```
    environment:
      - KOPIA_PASSWORD="P3pper61"
      - USER="elcid2012"
```

Because I like to use "-" lines in my YML files throughout, for consistency-sake; I find them easier to read.

By changing it to
```
    environment:
      - KOPIA_PASSWORD=P3pper61
      - USER=elcid2012
```
and rebooting the container, everything came back up as expected.

So now I'm wondering, what in the world is even the point of those environment variables? So, I brought the container down, deleted its persistent bind volume, and commented out those variables:
```
    environment:
#      - KOPIA_PASSWORD=P3pper61
#      - USER=elcid2012
```
Then brought the container back up (successful), recreated the S3/Minio repository (successful), and rebooted the container. Now I have this error:
```
ERROR unable to initialize repository: unable to initialize repository: error opening repository: get password: password prompt error: inappropriate ioctl for device
Enter password to open repository: 
Server will allow connections from users whose accounts are stored in the repository.
User accounts can be added using 'kopia server user add'.
```

But why? Why have those variables to begin with? For an S3/Minio respository, there's no username anyway! And the username/password combination to log into the Web GUI appears to be set with
```
    command:
      - server
      - start
      - --disable-csrf-token-checks
      - --insecure
      - --address=0.0.0.0:51515
      - --server-username=elcid2012
      - --server-password=P3pper61
```

So I don't understand what the point of them is. I think it's also especially egregious to keep either of those passwords in plain-text within the `docker-compose.yml` file.

If at any point the environment variable's `KOPIA_PASSWORD` is changed, then you lose access to the repository and you lose access to kopia.

**Comment by TilKenneth** [comment]
I am curious, can you somehow call `printenv` or similar right before server start? I don't know if you have to create a wrapper script, then you can see what the inside looks like just prior to start.

You would, say have a root process read a file, that is protected from all users, then this root process would start kopia passing along those environment variables, bypassing the command line arguments, running kopia as a low(er) privileged user.

There is
1. The GUI Server login.
2. The Server Control login.
3. The local repository password.
4. Then users inside the repository.

Ofcourse, the server part being a wrapper for another repository doesnt make it convenient, except from the user of that repository (KopiaUI).

**Comment by Cebrain** [comment]
I also have this error and it costs me some time to search the web, because i though i did something wrong.
I use docker on a linux server.

I created the container, setup the repo and everything works fine.
After reboot, then errors as specified in post 1.

After adding only the Variable "KOPIA_PASSWORD" with my Repo password the container starts again. ( i did not specify a user)
My Repo password are only numbers and letters and it is quite long.

Its strange that this problem is not addressed more, because it looks major to me.

I have 2 older Kopia instances, created months ago with an older Kopia versions and with them i did not specify any Variable, also now in updated latest version.

**Comment by audas** [comment]
Was this ever solved? I genuinely don't think Kopia is actually working in Docker after restarts. I have been through this process twice now - fully rebuilding everything deleting everything and it always fails to start up if the host machine or server goes down.

Any thoughts on this or just move onto another service - not very keen on being forced to install as native.

**Comment by audas** [comment]
Wonder if it has anything to do with this.

https://github.com/kopia/kopia/issues/4261

Either way - Kopia is non-functioning from Docker.

**Comment by elcid2012** [comment]
> Wonder if it has anything to do with this.
> 
> [#4261](https://github.com/kopia/kopia/issues/4261)
> 
> Either way - Kopia is non-functioning from Docker.

I gave up on kopia after this issue. Doesn’t seem to be properly configured that lends confidence as a safe backup solution. Plus the high number of open issues that just don’t seem to be getting looked at, and I would contribute but I don’t have the free time.

I am using backrest instead, which is a GUI front end for restic.

**Comment by audas** [comment]
Ok thanks. Was looking at restic or bacula. Will move on.

**Comment by vordenken** [comment]
Same problem for me. Why isn't this investigated further? Its a complete show-stopper.

**Comment by kntran1** [comment]
I wish I read this thread before setting up Kopia and backing up tons of data...  It worked really well, until the reboot...

How is something as simple as login authentication still an issue?  Looks like this was reported more than a year ago...

Anyway.. For those that are using docker, my workaround has been to delete the kopia config, then start it.  It boots up fresh and I just do the same repository setup and it finds everything, and I am able to restore.  Annoying but workaround for now until I find the time to setup something else..  Hopefully this gets fixed.

**Comment by Richardswe** [comment]
I agree with kntran1. The only thing i did was to add a couple of volumes in the docker-compose file and ran  "docker compose restart "  I will now build my backup solution with Ansible and Semaphore (i've done similiar before). 

" Server will allow connections from users whose accounts are stored in the repository.
User accounts can be added using 'kopia server user add'.

failed to open repository: unable to create format manager: invalid repository password
unable to initialize repository: unable to initialize repository: error opening repository: unable to open repository: unable to create format manager: invalid repository password

Server will allow connections from users whose accounts are stored in the repository.
User accounts can be added using 'kopia server user add'.
 "

**Comment by DalenW** [comment]
I figured it out, if this helps anyone else. You have to create the repository without setting that env variable. After that, then you need to set it for the container to auto-restart. I also didn't use any special characters in the repository password. Additionally, the environmental variable format in the docker compose file doesn't matter, I tried both to be sure.
